### PR TITLE
Migrate field refs in result_metadata from legacy format

### PR DESCRIPTION
**Pull request by calherries**
Epic: https://github.com/metabase/metabase/issues/30789
Pursuant towards https://github.com/metabase/metabase/issues/27504.

This PR adds a clojure migration to update legacy field refs in result_metadata. The migration is required for #27487, where we are matching field refs in `report_card.visualization_settings.column_settings` with field refs in `report_card.result_metadata`. `report_card.visualization_settings.column_settings` was updated in #30460 from legacy format, so for the match to work correctly `report_card.result_metadata` needs to be updated to.

Everything in this PR mirrors https://github.com/metabase/metabase/pull/30460, the only difference is that we are migrating `field_refs` in `result_metadata` instead of the keys of `visualization_settings.column_settings`.

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30787)

### CODE REVIEW

File: src/metabase/db/custom_migrations.clj [hunk 1]
```diff
@@ -173,7 +173,7 @@
                      :set    {:json_unfolding true}
                      :where  [:in :metabase_field.id field-ids-to-update]}))))
 
-(defn- update-legacy-field-refs [viz-settings]
```
Comment (calherries) [inline src/metabase/db/custom_migrations.clj:176]: I renamed this function used in the previous migration to differentiate it from `update-legacy-field-refs-in-result-metadata`.

File: src/metabase/db/custom_migrations.clj [hunk 2]
```diff
@@ -205,9 +205,40 @@
...
+                                    (let [updated (update-legacy-field-refs-in-viz-settings visualization_settings)]
                                       (when (not= visualization_settings updated)
                                         {:id                     id
                                          :visualization_settings updated}))))
                             (t2/reducible-query {:select [:id :visualization_settings]
                                                  :from   [:report_card]})))))
+
+(defn- update-legacy-field-refs-in-result-metadata [result-metadata]
+  (let [old-to-new (fn [ref]
+                     (match ref
```
Comment (calherries) [inline src/metabase/db/custom_migrations.clj:217]: Note this is exactly the same as the previous migration https://github.com/metabase/metabase/pull/30787/files#diff-be85115f22d8821f0a204e2c870538dc80f6c4191640e6f02c7cc558dbda6544R179

**Comment by qnkhuat** [comment]
the code change is good, but I'm worried doing migration on the card table and it could be very big, 

I posted a question in our data channel [here](https://metaboat.slack.com/archives/C01MC2CKW2G/p1684296407596899).

In the mean time maybe we could come up with some kind of profiling for this.

**Comment by calherries** [comment]
I just did a couple of test runs on my machine with postgres: 
- as a customer with 100,000 cards, updating all of them took 28.79 seconds.
- as a customer with 100,000 cards, updating none of them took 12.81 seconds.

These seem acceptable to me, I wouldn't be concerned unless it took more than a minute. And I suspect MySQL would not be 2x worse.

**Review by qnkhuat — APPROVED** [review]
The benchmark number LGTM.

**Pull request merged**